### PR TITLE
MON-3920: add runbook for PrometheusKubernetesListWatchFailures

### PR DESCRIPTION
**Pull request by machine424**

### CODE REVIEW

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+user workload monitoring is `openshift-user-workload-monitoring`.
+
+## Mitigation
+
+To gain further insight, review the logs of the affected Prometheus instance:
+
+   ```shell
+   $ NAMESPACE='<value of namespace label from alert>'
+
+   $ oc -n $NAMESPACE logs -c prometheus -l 'app.kubernetes.io/name=prometheus'
```
Comment (juzhao) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:28]: FYI, this command only output 20-line logs, maybe the error is not in the output, maybe
`$ oc -n $NAMESPACE logs -c prometheus ${prometheus_pod}`
is better?

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+```
+ts=2024-11-13T07:09:51.190Z caller=klog.go:108 level=warn component=k8s_client_runtime
+func=Warningf msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:554:
+failed to list *v1.Endpoints: endpoints is forbidden:
+User \"system:serviceaccount:openshift-monitoring:prometheus-k8s\" cannot list resource
+\"endpoints\" in API group \"\" in the namespace \"foo\""
+```
+
+To rectify this issue, ensure Prometheus is granted the necessary RBAC permissions as detailed in
+the following guide: [Configuring Prometheus to scrape metrics].
```
Comment (jan--f) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: Any reason to not just have the link here?

Comment (jan--f) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: yeah I get it. Not the biggest fan of working around it this way, but what can you do. I'll see if there is a lint exception. lgtm in the mean time.

Comment (simonpasquier) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: IIUC the permission issue can only happen for platform Prometheus and it's either a misconfiguration from the certified operator (product bug) or a user-defined service/pod monitor deployed in a plaform namespace (unsupported config). User-defined Prometheus should have full permissions by default.

The other cause could be a partial/complete outage of the Kubernetes API.

Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: It's legitimate markdown though that's already used in other places (https://github.com/search?q=repo%3Aopenshift%2Frunbooks+%22%5D%3A+%22&type=code), for me it's better than using an exception...

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 1 above)
Comment (juzhao) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:28]: I meant the command only output 20-line logs, the error maybe not in the 20-line logs
`oc -n $NAMESPACE logs -c prometheus -l 'app.kubernetes.io/name=prometheus' --tail=-1` will show all logs

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 2 above)
Comment (jan--f) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: Oh that actually gets rendered. TIL, sorry for the noise.

Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:49]: Yes, `User-defined Prometheus should have full permissions by default` but maybe there is a problem with deploying those needed RBAC, and not excluding UWM from the runbook may help debugging that.

This could also help SREs maybe on some setups (hypershift maybe? Not sure about that) where the default UWM RBAC are maybe tweaked (the default cluster role isn't used or sth).
I think we can keep it as generic as it is.

I'll try to reproduce and add the `Kube API outage` case.

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 1 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:28]: You're right, I didn't know about that, it's really hidden in the `--tail` description.
Some other commands with selectors in other runbooks would need to be adjusted.

**Comment by juzhao** [comment]
LGTM

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+The issue may arise for many reasons:
+
+### Insufficient RBAC Permissions for Prometheus
+
+In this scenario, Prometheus has been tasked to discover targets in a specified
+namespace, likely through `ServiceMonitor` or `PodMonitor` resources, yet
+Prometheus lacks the necessary RBAC permissions to query `Service`, `Endpoints`,
+`Pod`, and other related resources where the targets  are defined, logs like the
+following may be observed:
```
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:38]:
```suggestion
In this scenario, Prometheus has been tasked to discover targets in a specified
namespace, typically through `ServiceMonitor` or `PodMonitor` resources. However,
Prometheus does not have the necessary RBAC permissions to query `Service`, `Endpoints`,
`Pod`, and other related resources where the targets are defined. Logs similar to the
following one can be observed:
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+```text
+... failed to list *v1.Endpoints: endpoints is forbidden:
+User \"system:serviceaccount:openshift-monitoring:prometheus-k8s\" cannot list resource
+\"endpoints\" in API group \"\" in the namespace \"foo\""
+```
+
+To rectify this issue:
```
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:46]:
```suggestion
To resolve this issue, follow one of these steps:
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+- For user workload monitoring (`openshift-user-workload-monitoring`): Normally,
+Prometheus for user workload monitoring has cluster-wide permissions. Ensure
+that CMO has no issues deploying the underlying resources to obtain these
+permissions. Running `oc get co monitoring -o yaml` should provide hints.
+
+### Connectivity Issue
+
+Prometheus may not be able to reach the apiserver. In such cases, logs like the
+following may be seen (examples are not exhaustive):
```
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:60]:
```suggestion
Prometheus might not be able to reach the apiserver. In such cases, logs similar to the
following ones can be seen (examples are not exhaustive):
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+for default cluster monitoring or for the instance that monitors user-defined
+projects by viewing the alert message's `namespace` label: the namespace for
+default cluster monitoring is `openshift-monitoring` and the namespace for
+user workload monitoring is `openshift-user-workload-monitoring`.
+
+To gain further insight, review the logs of the affected Prometheus instance:
+
+   ```shell
+   # NAMESPACE='<value of namespace label from alert>'
+   $ oc -n  logs -c prometheus -l 'app.kubernetes.io/name=prometheus' --tail=-1
```
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:25]:
```suggestion
   $ oc -n $NAMESPACE logs -c prometheus -l 'app.kubernetes.io/name=prometheus' --tail=-1
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
+# PrometheusKubernetesListWatchFailures
+
+## Meaning
+
+The `PrometheusKubernetesListWatchFailures` alert is triggered when there is a constant
+increase in failures with `LIST/WATCH` requests to the Kubernetes API during Prometheus
```
[the remark this answers is not in the document]
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:6]: The former.

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 6 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:25]: nice catch

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (eromanova97) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:6]: Then,  for an additional clarity, I would rephrase it as follows: 

```suggestion
The `PrometheusKubernetesListWatchFailures` alert is triggered when `LIST/WATCH` requests to the Kubernetes API start failing during Prometheus
```

[the remark this answers is not in the document]
Comment (eromanova97) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:6]: But that is just a small suggestion, otherwise LGTM!

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+requests to the Kubernetes API start failing during Prometheus target discovery.
+
+## Impact
+
+This might prevent Prometheus from adding, updating, or deleting targets effectively.
+
+## Diagnosis
+
+Determine whether the alert has triggered for the instance of Prometheus used
+for default cluster monitoring or for the instance that monitors user-defined
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:15]:
```suggestion
for platform monitoring or for the instance that monitors user workloads
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+projects by viewing the alert message's `namespace` label. The namespace for
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:16]:
```suggestion
by checking the `namespace` label in  the alert message . The namespace for
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+default cluster monitoring is `openshift-monitoring` and the namespace for
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:17]:
```suggestion
platform monitoring is `openshift-monitoring` and the namespace for
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+```text
+... failed to list *v1.Endpoints: endpoints is forbidden:
+User \"system:serviceaccount:openshift-monitoring:prometheus-k8s\" cannot list resource
+\"endpoints\" in API group \"\" in the namespace \"foo\""
+```
+
+To resolve this issue, follow one of these steps:
+
+- For default cluster monitoring (`openshift-monitoring`): Follow
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:47]:
```suggestion
- For platform monitoring (`openshift-monitoring`): Follow
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+[Configuring Prometheus to scrape metrics] to ensure Prometheus is granted
+the necessary RBAC permissions.
+
+- For user workload monitoring (`openshift-user-workload-monitoring`): Typically,
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:51]:
```suggestion
- For user workload monitoring (`openshift-user-workload-monitoring`): Typically
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+Prometheus for user workload monitoring has cluster-wide permissions. Ensure
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:52]:
```suggestion
Prometheus deployed for user workload monitoring has cluster-wide permissions. Ensure
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+that CMO has no issues deploying the underlying resources to obtain these
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:53]:
```suggestion
that cluster-monitoring-operator (CMO) has no issues while deploying the underlying resources to obtain these
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+dial tcp A.B.C.D:443: i/o timeout"
+```
+
+```text
+... failed to list *v1.Endpoints: Get
+\"https://foo-apiserver:6443/api/v1/namespaces/openshift-console/endpoints?limit=500&
+resourceVersion=0\": dial tcp: lookup foo-apiserver on A.B.C.D:53: no such host"
+```
+
+Please refer to the relevant resources for debugging such problems.
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:73]: I think we should clarify here which resources to refer for more clarity

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+---
+
+If you cannot resolve the issue, log in to the
+[Customer Portal](https://access.redhat.com) and open a support case,
+attaching the artifacts gathered during the diagnosis procedure.
```
Comment (slashpai) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:79]: nit: adding the link to diagnosis procedure would be good to make it clear

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 8 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:15]: we use those wordings elsewhere https://github.com/search?q=repo%3Aopenshift%2Frunbooks%20%22default%20cluster%20monitoring%22&type=code and https://github.com/search?q=repo%3Aopenshift%2Frunbooks+%22user-defined%22&type=code
(better stay consistent as long as nobody complains about them)

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 10 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:17]: https://github.com/openshift/runbooks/pull/225#discussion_r1973370789

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 11 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:47]: https://github.com/openshift/runbooks/pull/225#discussion_r1973370789

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 16 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:79]: this same wording is used in many other places

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 15 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:73]: I wanted to have that as a placeholder for relevant links to be added. I'll remove it

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 9 above)
Comment (machine424) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:16]: same wording uses in other runbooks https://github.com/search?q=repo%3Aopenshift%2Frunbooks+%22projects+by+viewing+the+alert+message%22&type=code

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 16 above)
Comment (slashpai) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:79]: Agree, but I felt it's more clear if we provide the link correctly. But that can be done later for all runbooks

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 8 above)
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:15]: For slightly easier use of words:
```suggestion
Determine if the alert has triggered for the instance of Prometheus used
for default cluster monitoring or for the instance that monitors user-defined
```

Comment (eromanova97) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:15]: To put in my two cents,  in the official docs we would use "default platform" or "platform", but if nobody complained, it is not a big issue, it is still clear what we mean here :+1:

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 11 above)
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:47]:
```suggestion
- For default cluster monitoring (`openshift-monitoring`), follow
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 12 above)
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:51]:
```suggestion
- For user workload monitoring (`openshift-user-workload-monitoring`), typically,
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 14 above)
Comment (eromanova97) [reply alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:53]: To stay consistent with the official docs: "Cluster Monitoring Operator (CMO)"

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+the necessary RBAC permissions.
+
+- For user workload monitoring (`openshift-user-workload-monitoring`): Typically,
+Prometheus for user workload monitoring has cluster-wide permissions. Ensure
+that CMO has no issues deploying the underlying resources to obtain these
+permissions. Running the `oc get co monitoring -o yaml` command should provide hints.
+
+### Connectivity Issue
+
+Prometheus might not be able to reach the apiserver. In such cases, logs similar
```
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:58]:
```suggestion
Prometheus might not be able to reach the API server. In such cases, logs similar
```

File: alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md (hunk 15 above)
Comment (eromanova97) [inline alerts/cluster-monitoring-operator/PrometheusKubernetesListWatchFailures.md:73]:
```suggestion
Refer to the relevant resources to debug such problems.
```

**Review by eromanova97 — COMMENTED** [review]
Hello! I added just a few  more small suggestions I could think of, otherwise LGTM1

**Review by slashpai — COMMENTED** [review]
/lgtm
